bootloader/grub.go

```diff
@@ -379,11 +386,103 @@ func (g *grub) ManagedAssets() []string {
 	}
 }
 
-// CommandLine returns the kernel command line composed of the built-in
-// list and extra arguments passed in arguments. The command line may be
-// different when using a bootloader in the recovery partition.
+// CommandLine returns the kernel command line composed of system run mode
+// arguments, extra arguments (configured by snapd) and the built-in bootloader
+// specific arguments. The command line may be different when using a bootloader
+// in the recovery partition.
 //
 // Implements ManagedAssetsBootloader for the grub bootloader.
-func (g *grub) CommandLine(extra []string) (string, error) {
-	return "", fmt.Errorf("not implemented")
+func (g *grub) CommandLine(modeArgs, extraArgs string) (string, error) {
+	// we do not trust the on disk asset, use the built-in one
+	assetName := "grub.cfg"
+	isRecovery := false
+	if g.recovery {
+		assetName = "grub-recovery.cfg"
+		isRecovery = true
+	}
+	staticCmdline, err := staticCommandLineFromGrubAsset(assetName)
+	if err != nil {
+		return "", fmt.Errorf("cannot extract static command line element: %v", err)
+	}
+	// sort arguments so that they match their positions
```

Comment:
Silly question, why do we care about the sorting? Shouldn't the order be fixed from the grub assets?

Reply:
There's certain order fixed in grub config assets, `<mode> <static> <extra>`, where mode args are `snapd_recovery_mode=run` in grub.cfg and `snapd_recovery_mode=<recover|install> snapd_recovery_system=<label>` in grub-recovery.cfg. However, there's no rule that the caller of `bootloader.CommandLine()` must pass mode args in a specific order. At the same time we don't want to have an implicit assumption about the order of those arguments outside of a grub wrapper, which I think would be an implementation detail of grub config leaking up the stack.